### ISSUE TITLE
Unicode error when writing Species_annotation.txt

### ISSUE THREAD

**Issue by EarlyEvol** [nextgenusfs/funannotate#470]
TLDR: I have a unicode error when annotate is parsing the genbank file to print the annotation.txt file with library.py. The error is reproducible and originates from one entry.
```
  File "/home/u14/earlm1/miniconda3/envs/funannotate/lib/python2.7/site-packages/funannotate/library.py", line 7136, in annotationtable
    outfile.write('%s\n' % '\t'.join(result))
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 191: ordinal not in range(128)
```
I'm running funannotate 1.7.4 but an older build. I could update, but have run this pipe with a couple other genomes and would like to be consistent.
```funannotate    1.7.4-py27_0 --> 1.7.4-py27h864c0ab_1```
Command: 
```
funannotate annotate -i $funanno_out --cpus $threads \
--iprscan ./IPS/*.proteins.fa.xml \
--busco_db hymenoptera 
```
Error:
```
[02:57 PM]: Creating AGP file and corresponding contigs file
[02:57 PM]: Writing genome annotation table.
-------------------------------------------------------
Traceback (most recent call last):
  File "/home/u14/earlm1/miniconda3/envs/funannotate/bin/funannotate", line 668, in <module>
    main()
  File "/home/u14/earlm1/miniconda3/envs/funannotate/bin/funannotate", line 658, in main
    mod.main(arguments)
  File "/home/u14/earlm1/miniconda3/envs/funannotate/lib/python2.7/site-packages/funannotate/annotate.py", line 1448, in main
    final_annotation)
  File "/home/u14/earlm1/miniconda3/envs/funannotate/lib/python2.7/site-packages/funannotate/library.py", line 7136, in annotationtable
    outfile.write('%s\n' % '\t'.join(result))
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 191: ordinal not in range(128)
```
The error is reproducible after removing annotate_misc and annotate_results, so it isnt a one-off file corruption or something.

The offending entry is from the Interproscan xml. When I chop it out, annotate finishes with no error. 
```
<protein>
    <sequence md5="3cc68c9a8e9dbafa3e3e70f99e8c0df2">MFGENRCLPKFEDYIVSKILDLSNLVKLGSGICVKSSDVDMFTLSFFLWTVLRSIIHEIDTENKKELPLVLDDERRNSRRDSKENAIVLPQYDRRRPRVRRHSRRYYIERHRSFDVENGASTLDGGAGGGGGASPNAGLVLQNLPQRRESFLYRSDSDFEMSPKSMSRNSSIASERFKETEPPVERAHGEELIVTPFAQILASLRSVRNNFLSLTNVPTNKSRRSSGAQGSSTPQPRNLAPGEEPYMKMAVDTMEELDWCLDQLETIQTHRSVSDMASLKVRDAC</sequence>
    <xref id="GhFlFUN_023453-T1" name="GhFlFUN_023453-T1 GhFlFUN_023453"/>
    <matches>
      <hmmer3-match evalue="2.0E-31" score="109.2">
        <signature ac="PF18100" desc="Phosphodiesterase 4 upstream conserved regions (UCR)" name="PDE4_UCR">
          <entry ac="IPR040844" desc="Phosphodiesterase 4 upstream conserved regions (UCR)" name="PDE4_UCR" type="DOMAIN">
            <pathway-xref db="KEGG" id="00230+3.1.4.53" name="Purine metabolism"/>
            <pathway-xref db="Reactome" id="R-HSA-180024" name="DARPP-32 events"/>
          </entry>
          <signature-library-release library="PFAM" version="33.1"/>
        </signature>
        <model-ac>PF18100</model-ac>
        <locations>
          <hmmer3-location env-end="284" env-start="194" post-processed="true" score="108.2" evalue="4.0E-31" hmm-start="1" hmm-end="89" hmm-length="119" hmm-bounds="N_TERMINAL_COMPLETE" start="194" end="282">
            <location-fragments>
              <hmmer3-location-fragment start="194" end="282" dc-status="CONTINUOUS"/>
            </location-fragments>
          </hmmer3-location>
        </locations>
      </hmmer3-match>
      <mobidblite-match>
        <signature ac="mobidb-lite" desc="consensus disorder prediction" name="disorder_prediction">
          <signature-library-release library="MOBIDB_LITE" version="2.0"/>
        </signature>
        <model-ac>mobidb-lite</model-ac>
        <locations>
          <mobidblite-location sequence-feature="" start="216" end="243">
            <location-fragments>
              <mobidblite-location-fragment start="216" end="243" dc-status="CONTINUOUS"/>
            </location-fragments>
          </mobidblite-location>
        </locations>
      </mobidblite-match>
      <mobidblite-match>
        <signature ac="mobidb-lite" desc="consensus disorder prediction" name="disorder_prediction">
          <signature-library-release library="MOBIDB_LITE" version="2.0"/>
        </signature>
        <model-ac>mobidb-lite</model-ac>
        <locations>
          <mobidblite-location sequence-feature="" start="159" end="184">
            <location-fragments>
              <mobidblite-location-fragment start="159" end="184" dc-status="CONTINUOUS"/>
            </location-fragments>
          </mobidblite-location>
        </locations>
      </mobidblite-match>
      <mobidblite-match>
        <signature ac="mobidb-lite" desc="consensus disorder prediction" name="disorder_prediction">
          <signature-library-release library="MOBIDB_LITE" version="2.0"/>
        </signature>
        <model-ac>mobidb-lite</model-ac>
        <locations>
          <mobidblite-location sequence-feature="Polar" start="216" end="238">
            <location-fragments>
              <mobidblite-location-fragment start="216" end="238" dc-status="CONTINUOUS"/>
            </location-fragments>
          </mobidblite-location>
        </locations>
      </mobidblite-match>
      <panther-match evalue="4.7E-53" familyName="CAMP-SPECIFIC 3',5'-CYCLIC PHOSPHODIESTERASE 4B" score="183.6">
        <signature ac="PTHR11347:SF108" name="CAMP-SPECIFIC 3',5'-CYCLIC PHOSPHODIESTERASE 4B">
          <entry ac="IPR043371" desc="cAMP-specific 3’,5’-cyclic phosphodiesterase 4" name="PDE4" type="FAMILY">
            <go-xref category="MOLECULAR_FUNCTION" db="GO" id="GO:0004115" name="3',5'-cyclic-AMP phosphodiesterase activity"/>
            <go-xref category="BIOLOGICAL_PROCESS" db="GO" id="GO:0006198" name="cAMP catabolic process"/>
            <pathway-xref db="Reactome" id="R-HSA-180024" name="DARPP-32 events"/>
            <pathway-xref db="KEGG" id="00230+3.1.4.53" name="Purine metabolism"/>
          </entry>
          <signature-library-release library="PANTHER" version="14.1"/>
        </signature>
        <model-ac>PTHR11347:SF108</model-ac>
        <locations>
          <panther-location env-start="31" env-end="284" hmm-start="76" hmm-end="243" hmm-length="729" hmm-bounds="INCOMPLETE" start="104" end="281">
            <location-fragments>
              <panther-location-fragment start="104" end="281" dc-status="CONTINUOUS"/>
            </location-fragments>
          </panther-location>
        </locations>
      </panther-match>
      <panther-match evalue="4.7E-53" familyName="CAMP-SPECIFIC 3',5'-CYCLIC PHOSPHODIESTERASE 4B" score="183.6">
        <signature ac="PTHR11347" name="CYCLIC NUCLEOTIDE PHOSPHODIESTERASE">
          <signature-library-release library="PANTHER" version="14.1"/>
        </signature>
        <model-ac>PTHR11347</model-ac>
        <locations>
          <panther-location env-start="31" env-end="284" hmm-start="76" hmm-end="243" hmm-length="729" hmm-bounds="INCOMPLETE" start="104" end="281">
            <location-fragments>
              <panther-location-fragment start="104" end="281" dc-status="CONTINUOUS"/>
            </location-fragments>
          </panther-location>
        </locations>
      </panther-match>
    </matches>
  </protein>
```
All this info makes it through most of the pipe to create the genbank entry:
```
     gene            complement(29223422..29285879)
                     /locus_tag="GhFlFUN_023453"
     mRNA            complement(join(29223422..29223552,29225373..29225437,
                     29227024..29227125,29256160..29256192,29260776..29260964,
                     29283812..29284103,29285834..29285879))
                     /locus_tag="GhFlFUN_023453"
                     /product="hypothetical protein"
     CDS             complement(join(29223422..29223552,29225373..29225437,
                     29227024..29227125,29256160..29256192,29260776..29260964,
                     29283812..29284103,29285834..29285879))
                     /locus_tag="GhFlFUN_023453"
                     /note="GO_function: GO:0004115 - 3',5'-cyclic-AMP
                     phosphodiesterase activity [Evidence IEA];
                     GO_process: GO:0006198 - cAMP catabolic process [Evidence
                     IEA]"
                     /codon_start=1
                     /product="hypothetical protein"
                     /protein_id="ncbi:GhFlFUN_023453-T1"
                     /db_xref="InterPro:IPR040844"
                     /db_xref="InterPro:IPR043371"
                     /db_xref="PFAM:PF18100"
                     /translation="MFGENRCLPKFEDYIVSKILDLSNLVKLGSGICVKSSDVDMFTL
                     SFFLWTVLRSIIHEIDTENKKELPLVLDDERRNSRRDSKENAIVLPQYDRRRPRVRRH
                     SRRYYIERHRSFDVENGASTLDGGAGGGGGASPNAGLVLQNLPQRRESFLYRSDSDFE
                     MSPKSMSRNSSIASERFKETEPPVERAHGEELIVTPFAQILASLRSVRNNFLSLTNVP
                     TNKSRRSSGAQGSSTPQPRNLAPGEEPYMKMAVDTMEELDWCLDQLETIQTHRSVSDM
                     ASLKVRDAC"
```
Everything looks fine in this genbank entry as far as I can tell. 
I can just manually put in the couple of IPScan annotations, so it isnt a huge deal, but it is really bugging the crap out of me. If you have an idea of what the problem is let me know. If you think the new build will fix it, I'll just install it and redo the annotation steps for my other 2 genomes.

Thanks,
Earl

**Comment by nextgenusfs** [comment]
Hi Earl.  Sorry.  Py2 unicode errors are one of the most annoying..... The newest code in master is py3 compatible, so it will hopefully not be an issue anymore.  You could upgrade your condo environment to python3 and then install the latest from master, ie `python -m pip install git+https://github.com/nextgenusfs/funannotate.git`. 

But if you don't want to upgrade, you can probably fix by parsing/fixing the XML file -- its literally probably a single character causing the error (so annoying). I think you could probably just open the XML file in python and strip unicode characters like they do here: https://stackoverflow.com/a/25402141/4386003

maybe something like this (untested):
```
with open('fixed.xml', 'w') as outfile:
    with open('interpro.xml', 'r') as infile:
        for line in infile:
            line = line.encode('ascii', 'ignore').decode('ascii')
            outfile.write(line)
```

**Comment by EarlyEvol** [comment]
Jon,
Thanks for the help. It has set me on the right path. Using your little py script finds an error around a single (curly) right quote in the interpro xml. Not sure why it still throws an error instead of ignoring it, but I'm happy it did. Don't know why I didnt just look up \u2019 from the original error...which is that character. 
So I cleaned the interproscan file with sed
```cat interpro.xml | sed "s/’/'/g" > interpro.fix.xml```
Then the error persisted. I searched all the annotate_misc and annotate_result files and couldn't find an instance of the character, however I still got the same specific unicode error.
```UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 191: ordinal not in range(128)```
Went on a wild goose chase through the library.py and annotate.py scripts.  Finally found that annotate.py parses the $FUNANNOTATE_DB interpro file and passes it to annotationtable from the library. (I figured the descriptions were pulled from my interpro output file directly)
Of course the database is where my interpro scan output got the curly single quote. 
So! Just removed the curly single quotes from the interpro database and ran it again.
```cat $FUNANNOTATE_DB/interpro.xml | sed "s/’/'/g" > $FUNANNOTATE_DB/interpro.fix.xml```
That worked!

Not sure how this char got into the interpro database, or who to talk to to clean it up. I bet there are a bunch more hiding in there, so hopefully this helps someone else!

Thanks again!
Earl

**Comment by nextgenusfs** [comment]
Thanks Earl -- I hope this will no longer be a problem in version > 1.8 with python3... #fingerscrossed!
